### ISSUE TITLE
`MaskedTensor` 10x slower than `Tensor` compared with `nn.functional.softmax`

### ISSUE THREAD

**Issue by famura** [pytorch/pytorch#150186]
## Context

I was reading the docs on the new `MasedTensor` feature and wanted to get some insight if it might be useful in our transformer-based model. Thus, I started with pushing a demo input though the `softmax` function: once with a mask where certain values are set to `-inf` (typical for the attention mask in transformers), and once with the new `MasedTensor` feature. My hope was that the masked tensor would be faster as the masked values are skipped, but the opposite is the case.

Am I overlooking something? 

## MWE

Here is the code to reproduce my results

```py
import torch
from torch.nn.functional import softmax
from torch.masked import MaskedTensor
import torch.utils.benchmark as benchmark

# Define parameters
DEBUG = False
batch_size = 1 if DEBUG else 32
seq_len = 6 if DEBUG else 200
embed_dim = 4 if DEBUG else 64

# Set the seed
torch.manual_seed(0)

# Generate random input tensor
input_tensor = torch.randn(batch_size, seq_len, embed_dim)

# Generate a random mask
mask = torch.randint(0, 2, (batch_size, seq_len, embed_dim), dtype=torch.bool)

# Create the "normal" (tensor as it would be seen by our model)
input_with_minf = input_tensor.masked_fill(~mask, float("-inf"))

# Create masked tensor
masked_input = MaskedTensor(input_tensor, mask)

# Define the benchmarking tasks
t1 = benchmark.Timer(
    stmt="softmax(input_with_minf, dim=-1, dtype=torch.float32)",
    globals={"softmax": softmax, "input_with_minf": input_with_minf},
)

t2 = benchmark.Timer(
    stmt="softmax(masked_input, dim=-1, dtype=torch.float32)",
    globals={"softmax": softmax, "masked_input": masked_input},
)

# Run the benchmarks (run multiple iterations)
n_iter = 1 if DEBUG else 1000
normal_result = t1.timeit(n_iter)
masked_result = t2.timeit(n_iter) 

# Check if outputs are the same
softmax_output_normal = softmax(input_with_minf, dim=-1, dtype=torch.float32)
softmax_output_masked = softmax(masked_input, dim=-1, dtype=torch.float32)
assert isinstance(softmax_output_masked, MaskedTensor)
softmax_output_masked = softmax_output_masked.to_tensor(value=0.0) # 0.0 is softmax-specific
outputs_match = torch.allclose(softmax_output_normal, softmax_output_masked, equal_nan=True)

# Print results
print(f"Normal Tensor median time: {normal_result.median:.6f} s")
print(f"Masked Tensor median time: {masked_result.median:.6f} s")
print(f"Outputs Match: {outputs_match}")
```

## My Output (except the warnings about torch.masked being in prototype phase)

```sh
Normal Tensor median time: 0.000589 s
Masked Tensor median time: 0.006134 s
Outputs Match: True
```

**Comment by famura** [comment]
So when you mark things as `triaged`, what should I expect as a user?

**Comment by famura** [comment]
@malfet is there any hope that this issue will be addressed in the future?
